python/arcticdb/diagnostics_engine/version_list_diagnosis.py

```diff
@@ -0,0 +1,413 @@
+        )
+
+        node_colors.append(node_color)
+        node_labels[node_name] = node_label
+
+        # We need to mark the seen nodes to establish when a node is a tombstone and when it denotes the creation of a new version.
+        seen_nodes[current_key_version] = True
+
+        # Break loop if steps are 0 or current key version is target
+        if steps == 0 or current_key_version == target:
+            break
+
+        # Add edge to next node
+        if i < len(parsed_keys) - 1:
+            next_node_name = f"{parsed_keys[i + 1].version_id}_{i + 1}"
+            G.add_edge(node_name, next_node_name)
+
+    plt.figure(figsize=(12, 8))
+    pos = zigzag_layout(G)
+    nx.draw(G, pos, labels=node_labels, node_size=1200, arrows=True, node_color=node_colors)
+    add_graph_description(compacted_versions, exists_in_version_list, is_tombstoned)
+
+
+def parse_versions(versions):
+    """
+    Parse versions into a dictionary for easy access.
+    """
+    parsed_versions = {v["version"]: v for v in versions}
+    return parsed_versions
+
+
+def parse_keys(keys):
+    """
+    Parse and sort keys based on creation timestamp.
+    """
+    parsed_keys = sorted(keys, key=lambda k: k.creation_ts, reverse=True)
+    return parsed_keys
+
+
+def get_target_version_from_timestamp(parsed_versions, timestamp):
+    target_version = None
+    latest_timestamp = None
+    for version in parsed_versions.values():
+        if version["date"] < timestamp and not version["deleted"]:
+            if latest_timestamp is None or version["date"] > latest_timestamp:
+                latest_timestamp = version["date"]
+                target_version = version["version"]
+    return target_version
+
+
+def get_target_version(parsed_versions, parsed_keys, as_of):
+    """
+    Get the target version based on the 'as_of', or the latest version if 'as_of' is None.
+    """
+    if as_of is not None:
+        if isinstance(as_of, pd.Timestamp):
+            return get_target_version_from_timestamp(parsed_versions, as_of), False
+        elif isinstance(as_of, str):
+            return as_of, True
+        else:
+            return as_of, False
+
+    # When the target version is None, we need to return the most recent 'alive' version, which is the
+    # first non-tombstone one. The 'alive_versions' contains all the integers from the version list,
+    # but we must always respect the order established by 'parsed_keys' (i.e., the actual physical version chain).
+    # This is why we iterate across this strict order. Finally, we need to check the 'deleted' flag from
+    # the version list, as we still need to exclude the versions that are present in the version list
+    # solely because they are included in any snapshot.
+    alive_versions = [key for key in parsed_versions.keys() if parsed_versions[key]["deleted"] == False]
+    for key in parsed_keys:
+        if key.version_id in alive_versions:
+            return key.version_id, False
```

Comment:
I think this might miss the corner case where the latest live version is in a compacted key?
I think we should just use the output of `list_versions` to determine this though, live version numbers should always be increasing through the version chain.

Reply:
True